articles/libraries/auth0-android/save-and-refresh-tokens.md

```diff
@@ -0,0 +1,117 @@
+---
+section: libraries
+description: Keeping your user logged in with Auth0.Android
+---
+
+# Auth0.Android Saving and Renewing Tokens
+
+When an authentication is performed with the `offline_access` scope included, it will return a [refresh token](/refresh-token) that can be used to request a new user token, without authenticating the user again.
+
+## Credentials Manager
+
+[Auth0.Android](https://github.com/auth0/Auth0.Android) provides a utility class to streamline the process of storing and renewing credentials. You can access the `accessToken` or `idToken` properties from the [Credentials](https://github.com/auth0/Auth0.Android/blob/master/auth0/src/main/java/com/auth0/android/result/Credentials.java) instance. This is the preferred method to manage user credentials.
+
+First, add the library dependency to your build.gradle file:
+
+```gradle
+compile 'com.auth0.android:auth0:1.+'
+```
+
+Next, decide which class to use depending on your Android SDK target version:
+- If you require APIs lower than 21 (Lollipop) create a new instance of `CredentialsManager`. This class makes use of the `Context.MODE_PRIVATE` to store the credentials in a key-value preferences file, accessible only to your app. The data is stored in plaintext.
+- If you require APIs equal or greater than 21 (Lollipop) create a new instance of `SecureCredentialsManager`. This class has the same methods as the one above, but encrypts the data before storing it using a combination of _RSA_ and _AES_ algorithms along with the use of [Android KeyStore](https://developer.android.com/reference/java/security/KeyStore.html).
+
+
+## Using the CredentialsManager class
+
+### Setup
+Create a new instance by passing an `AuthenticationAPIClient` and a `Storage` implementation.
+
+```java
+Auth0 auth0 = new Auth0(this);
+AuthenticationAPIClient apiClient = new AuthenticationAPIClient(auth0);
+CredentialsManager manager = new CredentialsManager(apiClient, new SharedPreferencesStorage(this));
+```
+
+### Current State of the Authentication
+By asking for credentials existence upon launch and clearing them on logout, the manager can be used to manage the state of the authentication inside your app.
+
+Stored credentials are considered valid if they have not expired or can be refreshed. Check if a user has already logged in:
+
+```java
+boolean loggedIn = manager.hasValidCredentials();
+```
+
+When you want to log out the user from your app, remove the stored credentials and direct them to the login screen.
+
+```java
+manager.clearCredentials();
+//Show login screen
+```
+
+### Obtain Credentials
+Because the credentials may need to be refreshed against Auth0 Servers, this method is a-synchronic. Pass a callback implementation where you'd like to receive the credentials. Credentials returned by this method upon success are always valid.
```

Comment:
Do you need to talk about `a-synchronic` would just use `async` if you have to.
If it uses a callback isn't it implied to be async anyway?

Also didn't you suggest to me this should be called *Retrieving Credentials* 😄

Reply:
And this word, as far as I know, is `asynchronous` if you aren't just using `async`.

Reply:
I prefer asynchronous. Yes, it's implied but I think it's good to have a bit more text there just to intro the usage, don't you? I'll change it to "retrieving credentials".